### PR TITLE
bots: Generalize images to sharing of state

### PR DESCRIPTION
**Pull request by stefwalter**
The bots can share state. Previously this state was just created
images, but we generalize this to represent all state that
should be shared. This state is still in named files, just like
images.

 * [x] #7405

### CODE REVIEW

**Comment by stefwalter** [comment]
Tested like this:

```
$ bots/image-download --state centos-7-127.0.0.2-2202.log
 > https://209.132.184.69:8493/centos-7-127.0.0.2-2202.log
######################################################################## 100,0%
$ bots/image-upload --state centos-7-127.0.0.2-2202.log
######################################################################## 100,0%
$ bots/image-upload centos-7-127.0.0.2-2202.log
usage: image-upload [-h] [--store STORE] [--state] [image [image ...]]
image-upload: error: image link does not exist: centos-7-127.0.0.2-2202.log
$ bots/image-download centos-7-127.0.0.2-2202.log
image-download: image link does not exist: centos-7-127.0.0.2-2202.log
$ bots/image-download --state centos-7-127.0.0.2-2202.log
 > https://209.132.184.69:8493/centos-7-127.0.0.2-2202.log
######################################################################## 100,0%
$ bots/image-download --state centos-7-127.0.0.2-2202.log
 > https://209.132.184.69:8493/centos-7-127.0.0.2-2202.log

$ bots/image-download ubuntu-stable
 > https://209.132.184.69:8493/ubuntu-stable-8d82237ce28885c1feba0a48580ff09ba6f9a0e11ed74486ad79370b910540b1.qcow2
#################                                                         24,9%^C
$ bots/image-upload centos-7
#                                                                           1,1%^C
```

File: bots/image-upload [hunk 1]
```diff
@@ -59,17 +77,21 @@ def upload(store, source):
     return ret
 
 def main():
-    parser = argparse.ArgumentParser(description='Upload virtual machine images')
-    parser.add_argument("--store", default=UPLOAD, help="Where to send images")
+    parser = argparse.ArgumentParser(description='Upload bot state or images')
+    parser.add_argument("--store", default=UPLOAD, help="Where to send state or images")
+    parser.add_argument("--state", action="store_true", help="Loose images or state not recorded in git")
```
Comment (martinpitt) [inline bots/image-upload]: "Loose" sounds odd here - does that mean that they get removed locally? That would be weird for an image/state upload command.

File: bots/image-download [hunk 2]
```diff
@@ -141,48 +139,87 @@ def download(link, force, quiet, stores):
...
+
 
 def main():
-    parser = argparse.ArgumentParser(description='Download a virtual machine')
+    parser = argparse.ArgumentParser(description='Download a bot state or images')
     parser.add_argument("--force", action="store_true", help="Force unnecessary downloads")
-    parser.add_argument("--store", action="append", help="Where to find images")
+    parser.add_argument("--store", action="append", help="Where to find state or images")
     parser.add_argument("--quiet", action="store_true", help="Make downloading quieter")
+    parser.add_argument("--state", action="store_true", help="Loose state not recorded in git")
```
Comment (martinpitt) [inline bots/image-download]: Same here.

File: bots/image-download [hunk 3]
```diff
@@ -141,48 +139,87 @@ def download(link, force, quiet, stores):
...
+            shutil.move(temp, dest)
+
     finally:
         # if we had an error and the temp file is left over, delete it
         if os.path.exists(temp):
             os.unlink(temp)
 
-    # Handle alternate TEST_DATA
+# Calcalate a place to put images where links are not committed in git
+def state_target(image):
```
Comment (martinpitt) [inline bots/image-download]: Maybe call the argument `path`? `image` is a bit confusing here. (Also, calc*a*late typo in the comment, but unimportant)

File: bots/image-upload (hunk 1 above)
Comment (stefwalter) [reply bots/image-upload]: The Loose means they're not recorded in git. Lets just remove that word.

File: bots/image-download (hunk 3 above)
Comment (stefwalter) [reply bots/image-download]: We transform an image to a path here. I don't really care either way though. So changed.

File: bots/image-download (hunk 2 above)
Comment (stefwalter) [reply bots/image-download]: Changed here too.

**Comment by stefwalter** [comment]
Fixed for review.

**Pull request merged**